core/bin/migrations/20241203_fix_created_dsdocs.rs

```diff
@@ -0,0 +1,222 @@
+            "SELECT id, document_id, created, timestamp, tags_array, parents, source_url, hash, text_size, chunk_count, data_source \
+             FROM data_sources_documents \
+             WHERE created >= $1 AND created <= $2 AND id > $3 LIMIT $4",
+            &[&created_start, &created_end, &last_processed_id, &batch_size],
+        ).await?;
+
+        if dsdocs_rows.len() == 0 {
+            break;
+        }
+
+        last_processed_id = dsdocs_rows[dsdocs_rows.len() - 1].get(0);
+
+        println!("Getting data_sources");
+
+        let data_source_row_ids: Vec<i64> = dsdocs_rows.iter().map(|row| row.get(10)).collect();
+        let ds_rows = c
+            .query(
+                "SELECT id, data_source_id, project, internal_id FROM data_sources WHERE id = ANY($1)",
+                &[&data_source_row_ids],
+            )
+            .await?;
+
+        println!("Processing data_sources_documents");
+
+        #[derive(Debug, Clone)]
+        struct DsData {
+            data_source_id: String,
+            project: i64,
+            internal_id: String,
+        }
+
+        let ds_data: HashMap<i64, DsData> = ds_rows
+            .iter()
+            .map(|row| {
+                (
+                    row.get(0),
+                    DsData {
+                        data_source_id: row.get(1),
+                        project: row.get(2),
+                        internal_id: row.get(3),
+                    },
+                )
+            })
+            .collect();
+
+        // Tuples with (data_source_row_id, Document)
+        let docs: Vec<(DsData, i64, Document)> = dsdocs_rows
+            .iter()
+            .map(|row| {
+                let id: i64 = row.get(0);
+                let document_id: String = row.get(1);
+                let created: i64 = row.get(2);
+                let timestamp: i64 = row.get(3);
+                let tags_array: Vec<String> = row.get(4);
+                let parents: Vec<String> = row.get(5);
+                let source_url: Option<String> = row.get(6);
+                let hash: String = row.get(7);
+                let text_size: i64 = row.get(8);
+                let chunk_count: i64 = row.get(9);
+                let data_source_row_id: i64 = row.get(10);
+
+                let ds_data = ds_data
+                    .get(&data_source_row_id)
+                    .expect(&format!(
+                        "Unreachable: unknown data_source_row_id: {}",
+                        data_source_row_id
+                    ))
+                    .clone();
+                let data_source_id = ds_data.data_source_id.clone();
+
+                (
+                    ds_data,
+                    id,
+                    Document {
+                        data_source_id,
+                        created: created as u64,
+                        document_id,
+                        timestamp: timestamp as u64,
+                        tags: tags_array,
+                        parents,
```

Comment:
Why create a fully fledged document? vs just operate on docuemnt_id, hash  ?

Reply:
No good reason tbh

Reply:
fine with it but makes it a bit harder to understand

Reply:
fixed